File: ios/SkyCast/Features/Common/WeatherDetails.swift
import Foundation

/// Formats a ``Weather`` into the readings shown beneath the hero block.
///
/// Lives in `Features` rather than `Domain` because everything here is presentational: decimal
/// places, unit symbols, time format, and which shape the tile draws.
///
/// Labels are written inline, since on iOS a plain `String` literal is already a localisation key.
extension Weather {
    /// - Parameter includeDerived: adds the dew point. Off for Home, which stays a glance; on for
    ///   the detail screen, which is where someone goes precisely because the glance was not enough.
    func details(preferences: UserPreferences, includeDerived: Bool = false) -> [WeatherDetail] {
        let windUnit = preferences.windSpeedUnit
        let wind = windUnit.convertFromMetresPerSecond(windSpeedMetresPerSecond)
        // Beaufort is a force, not a speed: "5 Bft", never "5.0 Bft".
        let windText = windUnit.isWholeNumber
            ? "\(Int(wind.rounded())) \(windUnit.symbol)"
            : "\(wind.oneDecimalPlace) \(windUnit.symbol)"

        let pressureUnit = preferences.pressureUnit
        let pressure = pressureUnit.convertFromHectopascals(Double(pressureHpa))
        let pressureText = "\(pressure.formatted(places: pressureUnit.decimalPlaces)) \(pressureUnit.symbol)"

        let visibilityUnit = preferences.visibilityUnit
        let visibility = visibilityUnit.convertFromMetres(Double(visibilityMetres))

        return [
            WeatherDetail(
                label: "Humidity",
                value: "\(humidityPercent)%",
                kind: .humidity,
                // Humidity is already a percentage, so its fraction is itself.
                visual: .gauge(Double(humidityPercent) / 100)
            ),
            WeatherDetail(
                label: "Wind",
                value: windText,
                kind: .wind,
                // The direction has been in the model since the first commit and was never shown,
                // because a bar cannot draw a bearing. A compass can.
                visual: .compass(
                    degrees: Double(windDirectionDegrees),
                    cardinal: Self.cardinal(for: windDirectionDegrees)
                )
            ),
            WeatherDetail(
                label: "Pressure",
                value: pressureText,
                kind: .pressure,
                // Scaled across the range a barometer realistically covers, so the arc moves
                // meaningfully day to day.
                visual: .gauge((Double(pressureHpa) - Self.lowPressureHpa) / Self.pressureRangeHpa)
            ),
            WeatherDetail(
                label: "Visibility",
                value: "\(visibility.oneDecimalPlace) \(visibilityUnit.symbol)",
                kind: .visibility,
                // 10 km is the value OpenWeather reports for "clear", so it is effectively the top.
                visual: .gauge(Double(visibilityMetres) / Self.clearVisibilityMetres)
            ),
            WeatherDetail(
                label: "Cloud cover",
                value: "\(cloudinessPercent)%",
                kind: .cloud,
                visual: .gauge(Double(cloudinessPercent) / 100)
            ),
        ] + (includeDerived ? derivedDetails(preferences: preferences) : [])
    }

    private func derivedDetails(preferences: UserPreferences) -> [WeatherDetail] {
        let unit = preferences.temperatureUnit
        let dewPoint = unit.convertFromCelsius(dewPointCelsius)
        return [
            WeatherDetail(
                label: "Dew point",
                value: "\(Int(dewPoint.rounded()))\(unit.symbol)",
                kind: .dewPoint,
                // Measured against the air temperature: a dew point close to it is what "muggy"
                // actually means, and a nearly-full arc says so without the meteorology.
                visual: .gauge(dewPointCelsius / temperatureCelsius)
            ),
        ]
    }

    /// Everything the sun-path card needs, or `nil` where the times are unusable.
    func sunPath(now: Date = .now) -> SunPathReading? {
        let span = sunset.timeIntervalSince(sunrise)
        guard span > 0 else { return nil }
        return SunPathReading(
            progress: now.timeIntervalSince(sunrise) / span,
            sunriseLabel: localTime(sunrise),
            sunsetLabel: localTime(sunset),
            daylightLabel: daylightDuration.hoursAndMinutes,
            announcement: "Sunrise \(localTime(sunrise)), sunset \(localTime(sunset)), "
                + "\(daylightDuration.spokenDuration) of daylight"
        )
    }

    /// The 16-point compass name for a bearing, so "west-northwest" is available and not just
    /// "west".
    static func cardinal(for degrees: Int) -> String {
        let points = [
            "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
            "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW",
        ]
        // Normalised first: the API is documented as 0–360, but 360 and a negative are both
        // representable, and an out-of-range index here would crash the whole screen.
        let normalised = ((degrees % 360) + 360) % 360
        let sector = Int((Double(normalised) / 22.5).rounded()) % points.count
        return points[sector]
    }

    /// Formats an instant as a wall-clock time **in the observed location's zone**, not the
    /// device's: London's sunrise is 04:49 in London whatever the phone is set to.
    ///
    /// 24-hour clock, matching Android's `HH:mm`.
    private func localTime(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        formatter.timeZone = timeZone
        return formatter.string(from: date)
    }

    /// Roughly a strong gale: the point past which the wind indicator reads as full.
    private static let strongWindMetresPerSecond = 25.0
    private static let lowPressureHpa = 950.0
    private static let pressureRangeHpa = 130.0
    private static let clearVisibilityMetres = 10_000.0
}

/// The sun-path card's content, already formatted.
struct SunPathReading: Equatable {
    let progress: Double
    let sunriseLabel: String
    let sunsetLabel: String
    let daylightLabel: String
    let announcement: String
}

private extension TimeInterval {
    /// "14h 28m", rather than a count of seconds.
    var hoursAndMinutes: String {
        let minutes = Int(self / 60)
        return "\(minutes / 60)h \(minutes % 60)m"
    }

    /// The same duration in words, because "14h 28m" read aloud is not a sentence.
    var spokenDuration: String {
        let minutes = Int(self / 60)
        return "\(minutes / 60) hours \(minutes % 60) minutes"
    }
}

private extension Double {
    /// One decimal place, which is as much precision as these readings support.
    var oneDecimalPlace: Double {
        (self * 10).rounded() / 10
    }

    /// Formats to a fixed number of decimals. 0 gives a bare integer, not "1013.0".
    func formatted(places: Int) -> String {
        places == 0 ? "\(Int(rounded()))" : String(format: "%.\(places)f", self)
    }
}

File: android/app/src/main/java/com/nauhaan/skycast/ui/common/WeatherDetails.kt
package com.nauhaan.skycast.ui.common

import com.nauhaan.skycast.core.designsystem.component.WeatherDetail
import com.nauhaan.skycast.core.designsystem.component.WeatherDetailKind
import com.nauhaan.skycast.domain.model.UserPreferences
import com.nauhaan.skycast.domain.model.Weather
import java.time.Duration
import java.time.format.DateTimeFormatter
import kotlin.math.roundToInt

/**
 * Formats a [Weather] into the secondary readings shown under the hero block.
 *
 * Lives in `ui`, not `domain`, because everything here is presentational: decimal places, unit
 * symbols, time format. Labels arrive already resolved, so this stays free of Android's `Context`
 * and remains a plain testable function.
 *
 * @param includeDerived adds dew point and length of day. Off for Home, on for the detail screen.
 */
fun Weather.toDetails(
    preferences: UserPreferences,
    labels: WeatherDetailLabels,
    includeDerived: Boolean = false,
): List<WeatherDetail> {
    // The location's zone, not the device's: London's sunrise is 04:49 in London, and
    // reporting it as 09:49 because the phone is five hours ahead is simply wrong.
    val timeFormat = DateTimeFormatter.ofPattern(TIME_PATTERN).withZone(zoneOffset)

    val windUnit = preferences.windSpeedUnit
    val wind = windUnit.convertFromMetresPerSecond(windSpeedMetresPerSecond)
    val windText = if (windUnit.isWholeNumber) {
        // Beaufort is a force, not a speed: "5 Bft", never "5.0 Bft".
        "${wind.roundToInt()} ${windUnit.symbol}"
    } else {
        "${wind.toOneDecimalPlace()} ${windUnit.symbol}"
    }

    val pressureUnit = preferences.pressureUnit
    val pressure = pressureUnit.convertFromHectopascals(pressureHpa.toDouble())
    val pressureText = "${pressure.toPlaces(pressureUnit.decimalPlaces)} ${pressureUnit.symbol}"

    val visibilityUnit = preferences.visibilityUnit
    val visibility = visibilityUnit.convertFromMetres(visibilityMetres.toDouble())

    val derived = if (!includeDerived) {
        emptyList()
    } else {
        val temperatureUnit = preferences.temperatureUnit
        val dewPoint = temperatureUnit.convertFromCelsius(dewPointCelsius)
        listOf(
            WeatherDetail(
                label = labels.dewPoint,
                value = "${dewPoint.roundToInt()}${temperatureUnit.symbol}",
                kind = WeatherDetailKind.DEW_POINT,
                // Measured against the air temperature: a dew point close to it is what "muggy"
                // actually means, and a nearly-full bar says so without the meteorology.
                fraction = (dewPointCelsius / temperatureCelsius).coerceIn(0.0, 1.0),
            ),
            WeatherDetail(
                label = labels.daylight,
                value = daylightDuration.toHoursAndMinutes(),
                kind = WeatherDetailKind.DAYLIGHT,
                fraction = (daylightDuration.toMinutes() / MINUTES_IN_A_DAY).coerceIn(0.0, 1.0),
            ),
        )
    }

    return listOf(
        WeatherDetail(
            label = labels.humidity,
            value = "$humidityPercent%",
            kind = WeatherDetailKind.HUMIDITY,
            // Humidity is already a percentage, so its fraction is itself.
            fraction = humidityPercent / PERCENT,
        ),
        WeatherDetail(
            label = labels.wind,
            value = windText,
            kind = WeatherDetailKind.WIND,
            fraction = (windSpeedMetresPerSecond / STRONG_WIND_METRES_PER_SECOND).coerceIn(0.0, 1.0),
        ),
        WeatherDetail(
            label = labels.pressure,
            value = pressureText,
            kind = WeatherDetailKind.PRESSURE,
            // Scaled across the range a barometer realistically covers, so the indicator moves
            // meaningfully instead of sitting at the same spot every day.
            fraction = ((pressureHpa - LOW_PRESSURE_HPA) / PRESSURE_RANGE_HPA).coerceIn(0.0, 1.0),
        ),
        WeatherDetail(
            label = labels.visibility,
            value = "${visibility.toOneDecimalPlace()} ${visibilityUnit.symbol}",
            kind = WeatherDetailKind.VISIBILITY,
            // 10 km is the value OpenWeather reports for "clear", so it is effectively the ceiling.
            fraction = (visibilityMetres / CLEAR_VISIBILITY_METRES).coerceIn(0.0, 1.0),
        ),
        WeatherDetail(
            label = labels.sunrise,
            value = timeFormat.format(sunrise),
            kind = WeatherDetailKind.SUNRISE,
        ),
        WeatherDetail(
            label = labels.sunset,
            value = timeFormat.format(sunset),
            kind = WeatherDetailKind.SUNSET,
        ),
    ) + derived
}

/** "14h 28m", rather than a count of minutes. */
private fun Duration.toHoursAndMinutes(): String = "${toHours()}h ${toMinutesPart()}m"

/** One decimal place, which is as much precision as these readings support. */
private fun Double.toOneDecimalPlace(): Double = (this * ONE_DECIMAL_SCALE).roundToInt() / ONE_DECIMAL_SCALE

/** Formats to a fixed number of decimals. 0 gives a bare integer, not "1013.0". */
private fun Double.toPlaces(places: Int): String = if (places == 0) {
    roundToInt().toString()
} else {
    String.format(java.util.Locale.getDefault(), "%.${places}f", this)
}

private const val ONE_DECIMAL_SCALE = 10.0
private const val PERCENT = 100.0

/** Roughly a strong gale, the point past which the wind indicator should read as full. */
private const val STRONG_WIND_METRES_PER_SECOND = 25.0

private const val LOW_PRESSURE_HPA = 950.0
private const val PRESSURE_RANGE_HPA = 130.0
private const val CLEAR_VISIBILITY_METRES = 10_000.0

/** 24-hour clock: unambiguous, and the app is English-only for now. */
private const val TIME_PATTERN = "HH:mm"

/** The scale the daylight indicator reads against: a full 24 hours of sun. */
private const val MINUTES_IN_A_DAY = 24.0 * 60
